Worker panics when empty containerd config is present

## Short
When a empty containerd config file is present (E.g. everything is commented out) the worker node fails to startup.

## Proposal 

How to reproduce. 

1) My `/etc/containerd/config.toml`

```
#   Copyright 2018-2022 Docker Inc.

#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at

#       http://www.apache.org/licenses/LICENSE-2.0

#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

#disabled_plugins = ["cri"]

#root = "/var/lib/containerd"
#state = "/run/containerd"
#subreaper = true
#oom_score = 0

#[grpc]
#  address = "/run/containerd/containerd.sock"
#  uid = 0
#  gid = 0

#[debug]
#  address = "/run/containerd/debug.sock"
#  uid = 0
#  gid = 0
#  level = "info"
```

2) run the worker with `oak worker -d`

3) the logs ` sudo oak worker logs`
```
 sudo oak worker logs
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd3000c]

goroutine 1 [running]:
go_node_engine/virtualization.NewRuntimeManager.OnceValue[...].func2.1()
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/sync/oncefunc.go:54 +0x69
panic({0xef7540?, 0x19b17e0?})
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/runtime/panic.go:787 +0x132
go_node_engine/virtualization.newContainerdRuntime({{0xc0002803f0, 0xd}, {0xc0002bc720, 0x11}, {0xc0002bc768, 0x15}})
	/home/runner/work/oakestra/oakestra/go_node_engine/virtualization/ContainersManagement.go:96 +0x34c
go_node_engine/virtualization.NewRuntimeManager.func1()
	/home/runner/work/oakestra/oakestra/go_node_engine/virtualization/RuntimeDispatcher.go:44 +0x33
go_node_engine/virtualization.NewRuntimeManager.OnceValue[...].func2()
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/sync/oncefunc.go:57 +0x73
sync.(*Once).doSlow(0xffffffffffffffff?, 0x0?)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/sync/once.go:78 +0xab
sync.(*Once).Do(0xc00028041b?, 0xc0004242a0?)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/sync/once.go:69 +0x19
go_node_engine/virtualization.NewRuntimeManager.OnceValue[...].func3()
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.0.linux-amd64/src/sync/oncefunc.go:62 +0x36
go_node_engine/virtualization.(*RuntimeManager).GetRuntime(...)
	/home/runner/work/oakestra/oakestra/go_node_engine/virtualization/RuntimeDispatcher.go:55
main.main()
	/home/runner/work/oakestra/oakestra/go_node_engine/internal/daemon/nodeengined.go:56 +0x36f
^C
Detatching...
```

4) after setting the default containerd config  with
```
sudo containerd config default > /etc/containerd/config.toml
```
everything started working again

## Solution 
Improce in `ContainersManagement.go` the checks for the containerd config files when fetching additional oci runtimes. 

## Status
Looking for feedback

## Checklist
- [ ] Discussed
- [ ] Solved
- [ ] Tested

--- a/go_node_engine/virtualization/ContainersManagement.go
+++ b/go_node_engine/virtualization/ContainersManagement.go
@@ -37,11 +37,8 @@ import (
 
 func init() {
 	virtrt.Register(string(model.CONTAINER_RUNTIME), newContainerdRuntime)
-	plugins := findAdditionalRuntimePlugins()
-	if plugins != nil {
-		for additionalName := range findAdditionalRuntimePlugins() {
-			virtrt.Register(additionalName, newContainerdRuntime)
-		}
+	for additionalName := range findAdditionalRuntimePlugins() {
+		virtrt.Register(additionalName, newContainerdRuntime)
 	}
 }
 
@@ -104,17 +101,22 @@ func newContainerdRuntime(_ virtrt.RuntimeInfo) virtrt.Runtime {
 // Parses TOML directly: containerd's v2 LoadConfig rejects legacy short-form
 // disabled_plugins entries (e.g. "cri") that ship with Docker's containerd.
 func findAdditionalRuntimePlugins() iter.Seq[string] {
-	data, err := os.ReadFile(CONTAINERD_CONFIG_PATH)
+	return findAdditionalRuntimePluginsAt(CONTAINERD_CONFIG_PATH)
+}
+
+func findAdditionalRuntimePluginsAt(path string) iter.Seq[string] {
+	empty := maps.Keys(map[string]interface{}{})
+	data, err := os.ReadFile(path)
 	if err != nil {
 		logger.ErrorLogger().Printf("Unable to read containerd config file: %v", err)
-		return nil
+		return empty
 	}
 	var containerdConfig struct {
 		Plugins map[string]interface{} `toml:"plugins"`
 	}
 	if err := toml.Unmarshal(data, &containerdConfig); err != nil {
 		logger.ErrorLogger().Printf("Unable to parse containerd config file: %v", err)
-		return nil
+		return empty
 	}
 	for _, ctd := range containerdConfig.Plugins {
 		ctd, ok := ctd.(map[string]interface{})["containerd"].(map[string]interface{})
@@ -122,13 +124,13 @@ func findAdditionalRuntimePlugins() iter.Seq[string] {
 			runtimes, ok := ctd["runtimes"].(map[string]interface{})
 			if ok {
 				for runtimeName := range runtimes {
-					logger.InfoLogger().Printf("Adding compatibility custom runtime %s configured in containerd config file %s", runtimeName, CONTAINERD_CONFIG_PATH)
+					logger.InfoLogger().Printf("Adding compatibility custom runtime %s configured in containerd config file %s", runtimeName, path)
 				}
 				return maps.Keys(runtimes)
 			}
 		}
 	}
-	return nil
+	return empty
 }
 
 // StopContainerdClient stops the container runtime client

--- a/go_node_engine/virtualization/ContainersManagement.go
+++ b/go_node_engine/virtualization/ContainersManagement.go
@@ -90,8 +90,13 @@ func newContainerdRuntime(_ virtrt.RuntimeInfo) virtrt.Runtime {
 	runtime.mountedVolumes = make(map[string][]csi.MountedVolume)
 
 	// Advertise sub-runtimes discovered from the containerd config (e.g. runc).
+	found := false
 	for name := range findAdditionalRuntimePlugins() {
 		model.GetNodeInfo().AddSupportedTechnology(model.RuntimeType(name))
+		found = true
+	}
+	if !found {
+		logger.WarnLogger().Printf("No additional OCI runtimes found in containerd config %s", CONTAINERD_CONFIG_PATH)
 	}
 
 	return &runtime
@@ -126,6 +131,7 @@ func findAdditionalRuntimePlugins() iter.Seq[string] {
 			}
 		}
 	}
+	logger.WarnLogger().Printf("No OCI runtimes found in containerd config %s", CONTAINERD_CONFIG_PATH)
 	return empty
 }
 

--- a/go_node_engine/virtualization/ContainersManagement.go
+++ b/go_node_engine/virtualization/ContainersManagement.go
@@ -37,11 +37,8 @@ import (
 
 func init() {
 	virtrt.Register(string(model.CONTAINER_RUNTIME), newContainerdRuntime)
-	plugins := findAdditionalRuntimePlugins()
-	if plugins != nil {
-		for additionalName := range findAdditionalRuntimePlugins() {
-			virtrt.Register(additionalName, newContainerdRuntime)
-		}
+	for additionalName := range findAdditionalRuntimePlugins() {
+		virtrt.Register(additionalName, newContainerdRuntime)
 	}
 }
 
@@ -104,17 +101,18 @@ func newContainerdRuntime(_ virtrt.RuntimeInfo) virtrt.Runtime {
 // Parses TOML directly: containerd's v2 LoadConfig rejects legacy short-form
 // disabled_plugins entries (e.g. "cri") that ship with Docker's containerd.
 func findAdditionalRuntimePlugins() iter.Seq[string] {
+	empty := func(yield func(string) bool) {}
 	data, err := os.ReadFile(CONTAINERD_CONFIG_PATH)
 	if err != nil {
 		logger.ErrorLogger().Printf("Unable to read containerd config file: %v", err)
-		return nil
+		return empty
 	}
 	var containerdConfig struct {
 		Plugins map[string]interface{} `toml:"plugins"`
 	}
 	if err := toml.Unmarshal(data, &containerdConfig); err != nil {
 		logger.ErrorLogger().Printf("Unable to parse containerd config file: %v", err)
-		return nil
+		return empty
 	}
 	for _, ctd := range containerdConfig.Plugins {
 		ctd, ok := ctd.(map[string]interface{})["containerd"].(map[string]interface{})
@@ -128,7 +126,7 @@ func findAdditionalRuntimePlugins() iter.Seq[string] {
 			}
 		}
 	}
-	return nil
+	return empty
 }
 
 // StopContainerdClient stops the container runtime client

--- a/go_node_engine/virtualization/Containers_test.go
+++ b/go_node_engine/virtualization/Containers_test.go
@@ -1,6 +1,8 @@
 package virtualization
 
 import (
+	"os"
+	"path/filepath"
 	"testing"
 
 	"gotest.tools/v3/assert"
@@ -34,4 +36,56 @@ func TestInstanceExtractionFromTaskidMultipleInstance(t *testing.T) {
 	taskid := genTaskID(sname, iid)
 	extracted := extractInstanceNumberFromTaskID(taskid)
 	assert.Equal(t, extracted, iid)
+}
+
+// Regression test for #512: ranging the result of findAdditionalRuntimePlugins
+// must not panic when the containerd config file is missing, empty, or fully
+// commented out (all top-level keys absent).
+func TestFindAdditionalRuntimePluginsHandlesEmptyConfig(t *testing.T) {
+	cases := map[string]string{
+		"empty file":         "",
+		"fully commented":    "# disabled_plugins = [\"cri\"]\n# [grpc]\n# address = \"/run/containerd/containerd.sock\"\n",
+		"no plugins section": "version = 2\n",
+	}
+	for name, contents := range cases {
+		t.Run(name, func(t *testing.T) {
+			dir := t.TempDir()
+			path := filepath.Join(dir, "config.toml")
+			assert.NilError(t, os.WriteFile(path, []byte(contents), 0o644))
+
+			count := 0
+			for range findAdditionalRuntimePluginsAt(path) {
+				count++
+			}
+			assert.Equal(t, count, 0)
+		})
+	}
+}
+
+func TestFindAdditionalRuntimePluginsHandlesMissingFile(t *testing.T) {
+	missing := filepath.Join(t.TempDir(), "does-not-exist.toml")
+	count := 0
+	for range findAdditionalRuntimePluginsAt(missing) {
+		count++
+	}
+	assert.Equal(t, count, 0)
+}
+
+func TestFindAdditionalRuntimePluginsReturnsConfiguredRuntimes(t *testing.T) {
+	dir := t.TempDir()
+	path := filepath.Join(dir, "config.toml")
+	contents := `version = 2
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
+runtime_type = "io.containerd.runc.v2"
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.kata]
+runtime_type = "io.containerd.kata.v2"
+`
+	assert.NilError(t, os.WriteFile(path, []byte(contents), 0o644))
+
+	got := map[string]bool{}
+	for name := range findAdditionalRuntimePluginsAt(path) {
+		got[name] = true
+	}
+	assert.Assert(t, got["runc"], "expected runc, got %v", got)
+	assert.Assert(t, got["kata"], "expected kata, got %v", got)
 }

